File: sdk/python/captcha-background-sdk/captcha_background_sdk/text_layer_renderer.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Optional

from PIL import Image

from .types import BBox, TextRegion


@dataclass(frozen=True)
class TextLayerRenderOutput:
    text_bbox: Optional[BBox]
    text_pixel_count: int
    output_path: Optional[str]


def _to_pil_crop_box(bbox: BBox) -> tuple[int, int, int, int]:
    left, top, right, bottom = bbox
    return (left, top, right + 1, bottom + 1)


def _safe_pixel_index(width: int, height: int, x: int, y: int) -> Optional[int]:
    if x < 0 or y < 0 or x >= width or y >= height:
        return None
    return y * width + x


def render_text_layer(
    captcha_path: str,
    regions: Iterable[TextRegion],
    text_bbox: Optional[BBox],
    output_path: Optional[str],
    crop_to_content: bool,
) -> TextLayerRenderOutput:
    with Image.open(captcha_path) as image:
        src = image.convert("RGBA")
        width, height = src.size
        src_pixels = list(src.getdata())

    out_pixels = [(0, 0, 0, 0)] * (width * height)
    picked = set()
    for region in regions:
        for x, y in region.pixels:
            idx = _safe_pixel_index(width, height, x, y)
            if idx is None or idx in picked:
                continue
            picked.add(idx)
            out_pixels[idx] = src_pixels[idx]

    out_image = Image.new("RGBA", (width, height), (0, 0, 0, 0))
    out_image.putdata(out_pixels)

    if crop_to_content and text_bbox is not None:
        out_image = out_image.crop(_to_pil_crop_box(text_bbox))

    normalized_output_path: Optional[str] = None
    if output_path:
        target = Path(output_path)
        if target.parent and not target.parent.exists():
            target.parent.mkdir(parents=True, exist_ok=True)
        out_image.save(target)
        normalized_output_path = str(target)

    return TextLayerRenderOutput(
        text_bbox=text_bbox,
        text_pixel_count=len(picked),
        output_path=normalized_output_path,
    )

File: typescript-sdk/src/text-layer-renderer.ts
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';
import { BBox, TextRegion } from './types';

export interface TextLayerRenderOutput {
  text_bbox: BBox | null;
  text_pixel_count: number;
  output_path: string | null;
}

function toSharpBox(bbox: BBox) {
  const [left, top, right, bottom] = bbox;
  return { left, top, width: right - left + 1, height: bottom - top + 1 };
}

function safePixelIndex(width: number, height: number, x: number, y: number): number | null {
  if (x < 0 || y < 0 || x >= width || y >= height) return null;
  return y * width + x;
}

export async function renderTextLayer(
  captchaPath: string,
  regions: TextRegion[],
  textBBox: BBox | null,
  outputPath?: string,
  cropToContent: boolean = false
): Promise<TextLayerRenderOutput> {
  const { data, info } = await sharp(captchaPath).ensureAlpha().raw().toBuffer({ resolveWithObject: true });
  const width = info.width;
  const height = info.height;
  const channels = info.channels;
  if (!width || !height || !channels) {
    throw new Error('invalid_image_dimensions');
  }

  const out = Buffer.alloc(width * height * 4, 0);
  const picked = new Set<number>();

  for (const region of regions) {
    for (const [x, y] of region.pixels) {
      const idx = safePixelIndex(width, height, x, y);
      if (idx === null || picked.has(idx)) {
        continue;
      }
      picked.add(idx);
      const base = idx * channels;
      const outBase = idx * 4;
      out[outBase] = data[base];
      out[outBase + 1] = data[base + 1];
      out[outBase + 2] = data[base + 2];
      out[outBase + 3] = channels > 3 ? data[base + 3] : 255;
    }
  }

  let output = sharp(out, { raw: { width, height, channels: 4 } });
  if (cropToContent && textBBox) {
    output = output.extract(toSharpBox(textBBox));
  }

  let normalizedOutput: string | null = null;
  if (outputPath) {
    const target = path.resolve(outputPath);
    await fs.promises.mkdir(path.dirname(target), { recursive: true });
    await output.png().toFile(target);
    normalizedOutput = target;
  }

  return {
    text_bbox: textBBox,
    text_pixel_count: picked.size,
    output_path: normalizedOutput,
  };
}
